compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ServiceDesugar.java

```diff
@@ -187,13 +196,26 @@ void rewriteServiceVariable(BLangService service, SymbolEnv env, BLangBlockStmt
         }
     }
 
+    private BType getListenerType(BType type) {
+        if (type.tag == TypeTags.UNION) {
+            for (BType memberType : ((BUnionType) type).getMemberTypes()) {
+                if (types.checkListenerCompatibility(memberType)) {
+                    return memberType;
+                }
+            }
+        }
+        return type;
+    }
+
     private void addMethodInvocation(Location pos, BLangSimpleVarRef varRef, BInvokableSymbol methodRefSymbol,
                                      List<BLangExpression> args,
                                      BlockNode body) {
         // Create method invocation
         final BLangInvocation methodInvocation =
                 ASTBuilderUtil.createInvocationExprForMethod(pos, methodRefSymbol, args, symResolver);
-        methodInvocation.expr = varRef;
+        BLangCheckedExpr listenerCheckExpr = ASTBuilderUtil.createCheckExpr(pos, varRef, getListenerType(varRef.type));
```

Comment:
Since the listener type should be a subtype of listener, `check` needs to be added only to the listener initialization (and when the expression type includes error), right? 

Do we need this change?

Reply:
Adding check here wouldn't add it for the initialization anyway, right?

The following results in a bad, sad error.

```ballerina
public class Listener {
    public function init() returns error? => error("err");

    public isolated function 'start() returns error? { }

    public isolated function gracefulStop() returns error? { }

    public isolated function immediateStop() returns error? { }

    public isolated function detach(service object {} s) returns error? { }

    public isolated function attach(service object {} s, string[]|string? name = ()) returns error? { }
}

listener Listener ln =  new;
```

```log
[2021-01-14 23:10:56,517] SEVERE {b7a.log.crash} - class io.ballerina.runtime.internal.values.ErrorValue cannot be cast to class io.ballerina.runtime.api.values.BObject (io.ballerina.runtime.internal.values.ErrorValue and io.ballerina.runtime.api.values.BObject are in unnamed module of loader 'app') 
java.lang.ClassCastException: class io.ballerina.runtime.internal.values.ErrorValue cannot be cast to class io.ballerina.runtime.api.values.BObject (io.ballerina.runtime.internal.values.ErrorValue and io.ballerina.runtime.api.values.BObject are in unnamed module of loader 'app')
	at $$$.$gen$$0046$0046$0060init$00620(.:15)
	at $_init.$gen$$0046$0060init$0062(.:0)
	at $_init.$moduleInit(.)
	at $_init.$lambda$$moduleInit$(.)
	at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(Scheduler.java:571)
	at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:301)
	at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:269)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Reply:
Oh I totally forgot about the listener-decl 😞

Reply:
Fixed. And a test case is added https://github.com/ballerina-platform/ballerina-lang/pull/27605/files#diff-dedee6299f9c3f21babab9613444bee606906b9323bde0ac0cc0b0d215865d42R79